### PR TITLE
CC26X2X7 Initial support of BLE Rendezvous

### LINKED ISSUE #4864: Texas Instruments CC26X2 does not support BLE Rendezvous
#### Problem

The CC26X2 is a multi-band LPRF SoC, with the capability to run both IEEE 802.15.4 and BLE. This platform should support BLE Rendezvous according to the CHIP specification.

 #### Proposed Solution

The new CC26X2X7 SoC has sufficient Flash and RAM space to support concurrent Thread and BLE operation. The platform adaptation layer and example code should be updated to target this new platform and enable this operation.

### PR DESCRIPTION
**Pull request by srickardti**
Migrate to the new Texas Instruments CC26X2X7 LaunchPad. This allows
concurrent Full Thread Device and BLE operation. This Enables BLE
rendezvous on the CC26X2X7.

Thanks to Alexander D'Abreu for work on the DMM integration and CHIPoBLE
profile development for the TI platform.

Co-Authored-by: Alexander D'Abreu <alexander.dabreu@ti.com>

 #### Problem

The Texas Instruments CC26X2 platform does not enable BLE Rendezvous according to the CHIP specification.

 #### Summary of Changes
 
* Migrates the build to denote the new CC26X2X7 platform support
* Enables concurrent Thread and BLE operation for the CC26X2 platform.
* Adds CHIPoBLE profile for BLE Rendezvous for TI CC26X2 platform

 Fixes #4864

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.h
Comment (andy31415) [inline src/platform/cc13x2_26x2/BLEManagerImpl.h:251]: can we remove the commented out code?

**Review by andy31415 — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.h
Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.h]: Thought I got all of that. should be clean now

**Review by mspang — COMMENTED** [review]
The recommended board changed?

**Comment by srickardti** [comment]
@mspang Running the Thread stack and BLE stack together requires more Flash space than the CC2652R1 has available. The new CC2652R7 has 704K of flash to accommodate this requirement. Same platform with more Flash/RAM.

File: examples/lock-app/cc13x2x7_26x2x7/README.md
Comment (srickardti) [reply examples/lock-app/cc13x2x7_26x2x7/README.md:68]: That references the link on line 230 of the README

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:281]: mis-match of our coding conventions from BLE, I'll remove them.

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:188]: I'll update those, since the buffer size was statically allocated, we thought parameter would be a better fit.

**Review by woody-apple — APPROVED** [review]
Approving pending @mspang

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: We could return a `false`. I will add this instead.

Comment (mspang) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: Would assert() or nlASSERT() be appropriate here?

Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: I think we were unsure about what the best way forward is for these cases. For this one, returning a `false` is probably correct. There are others that the function does not have a return type, so we cannot propagate an error up. for that, would `assert()` be correct?

Comment (mspang) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: assert() should be preferred for any internal error or precondition failure. For public APIs, an error return may also be appropriate.

i.e., if the condition being false is not expected to happen unless there's a bug, we **should** assert.

Conversely, if an assert fires, it **should** be the case that there is a software bug or hardware failure

If an error condition happens that is **not** the result of a software bug, we must not assert.

Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: So for these functions, returning a `false` on an allocation failure makes sense.

On some of the others, like failing to allocate the BLE task, I could add the return value path or assert?

Comment (mspang) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:324]: I think it's reasonable to assert if we cannot allocate the task - it'd a bug if we hit this and not recoverable.

File: src/platform/BUILD.gn
Comment (mspang) [reply src/platform/BUILD.gn:255]: This needs fixing. I think you just need to  add `"${sdk_target_name}_posix_config"` to **public_configs** of `"${chip_root}/third_party/ti_simplelink_sdk"`  rather than **configs**.

**Review by mspang — CHANGES_REQUESTED** [review]
Request changes for the build file bit which should be fixed up.

**Comment by srickardti** [comment]
Hey, had to deal with some stuff AFK. The state of Texas isn't in the best state right now. I was looking into it and found the header that needed the TI-POSIX API was removed somewhere along the line of developing this feature. But the build configuration wasn't updated. I've removed the offending build config.

The issue with adding the TI-POSIX config to the `${chip_root}/third_party/ti_simplelink_sdk` target's public configs is that there isn't a way to propagate a negative requirement to the dependencies. The headers conflict with some of the GNU definitions pulled in by `-std=gnu++14` and `-std=gnu11`.

**Review by mspang — APPROVED** [review]

**Comment by mspang** [comment]
> Hey, had to deal with some stuff AFK. The state of Texas isn't in the best state right now. I was looking into it and found the header that needed the TI-POSIX API was removed somewhere along the line of developing this feature. But the build configuration wasn't updated. I've removed the offending build config.
> 
> The issue with adding the TI-POSIX config to the `${chip_root}/third_party/ti_simplelink_sdk` target's public configs is that there isn't a way to propagate a negative requirement to the dependencies. The headers conflict with some of the GNU definitions pulled in by `-std=gnu++14` and `-std=gnu11`.

Ah, right, got it. The main thing that needs fixing is just using // to refer to it instead of simplelink_sdk_root, chip_root, or similar. I don't want us to depend on where the code is in the checkout, in general, so we have the foo_root variables to avoid it.

You can put a config in third_party/simplelink and reference it via simplelink_root. Oe, for things that the application might replace, we have the option of using declare_arg() to point a group target at it. That's the only way we should be referring to an application provided library from common code such as src/platform.

**Comment by mspang** [comment]
> > Hey, had to deal with some stuff AFK. The state of Texas isn't in the best state right now. I was looking into it and found the header that needed the TI-POSIX API was removed somewhere along the line of developing this feature. But the build configuration wasn't updated. I've removed the offending build config.
> > The issue with adding the TI-POSIX config to the `${chip_root}/third_party/ti_simplelink_sdk` target's public configs is that there isn't a way to propagate a negative requirement to the dependencies. The headers conflict with some of the GNU definitions pulled in by `-std=gnu++14` and `-std=gnu11`.
> 
> Ah, right, got it. The main thing that needs fixing is just using // to refer to it instead of simplelink_sdk_root, chip_root, or similar. I don't want us to depend on where the code is in the checkout, in general, so we have the foo_root variables to avoid it.
> 
> You can put a config in third_party/simplelink and reference it via simplelink_root. Oe, for things that the application might replace, we have the option of using declare_arg() to point a group target at it. That's the only way we should be referring to an application provided library or config target from common code such as src/platform.

BTW, can we just replace gnu++14 with c++14 for the whole TI build, not on a target by target basis? If so, that's a good solution.

**Review by msandstedt — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (jelderton) [inline src/platform/cc13x2_26x2/BLEManagerImpl.cpp:329]: is this suppose to be `ICall_free((void *) pMsg);`

Comment (srickardti) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp:329]: good catch, yes it is.

**Review by jelderton — APPROVED** [review]

File: src/platform/cc13x2_26x2/BLEManagerImpl.cpp
Comment (jelderton) [reply src/platform/cc13x2_26x2/BLEManagerImpl.cpp]: Thanks for the quick response.  Waiting for all of the checks to complete, and I'll merge this in.

**Pull request merged**
